lib/Db/FeedMapperV2.php

```diff
@@ -158,23 +160,34 @@ public function findAllFromFolder(?int $id): array
      * @param string   $userId
      * @param int      $id
      * @param int|null $maxItemID
+     *
+     * @throws \OCP\DB\Exception
+     *
+     * @TODO Update for NC 21
      */
     public function read(string $userId, int $id, ?int $maxItemID = null): void
     {
-        $builder = $this->db->getQueryBuilder();
-        $builder->update(ItemMapperV2::TABLE_NAME, 'items')
+        $idBuilder = $this->db->getQueryBuilder();
+        $idBuilder->select('items.id')
+            ->from(ItemMapperV2::TABLE_NAME, 'items')
             ->innerJoin('items', FeedMapperV2::TABLE_NAME, 'feeds', 'items.feed_id = feeds.id')
-            ->setValue('unread', 0)
             ->andWhere('feeds.user_id = :userId')
-            ->andWhere('feeds.id = :feedId')
-            ->setParameter('userId', $userId)
-            ->setParameter('feedId', $id);
+            ->andWhere('feeds.id = :feedId');
+
+        $idList = $this->db->executeQuery($idBuilder->getSQL(), $idBuilder->getParameters())
+                           ->fetch(PDO::FETCH_COLUMN);
```

Comment:
```php
            ->setParameter('userId', $userId)
            ->setParameter('feedId', $id);
```
is still required if you use `$idBuilder->getParameters()`. Don't mix-up my two suggestions from last time 🙂.
<details><summary>Related log entry</summary>

```json
{
    "reqId": "9soKnvrwj6elLQBXQejk",
    "level": 3,
    "time": "2021-03-20T17:10:41+00:00",
    "remoteAddr": "127.0.0.1",
    "user": "anoymouserver",
    "app": "index",
    "method": "POST",
    "url": "/apps/news/feeds/135/read",
    "message": {
        "Exception": "Doctrine\\DBAL\\Exception\\SyntaxErrorException",
        "Message": "An exception occurred while executing 'SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)':\n\nSQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1",
        "Code": 0,
        "Trace": [{
                "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php",
                "line": 169,
                "function": "convertException",
                "class": "Doctrine\\DBAL\\Driver\\AbstractMySQLDriver",
                "type": "->",
                "args": ["An exception occurred while executing 'SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)':\n\nSQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1", {
                        "errorInfo": ["42000", 1064, "You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1"],
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOException"
                    }
                ]
            }, {
                "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php",
                "line": 149,
                "function": "wrapException",
                "class": "Doctrine\\DBAL\\DBALException",
                "type": "::",
                "args": [{
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOMySql\\Driver"
                    }, {
                        "errorInfo": ["42000", 1064, "You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1"],
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOException"
                    }, "An exception occurred while executing 'SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)':\n\nSQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1"]
            }, {
                "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php",
                "line": 914,
                "function": "driverExceptionDuringQuery",
                "class": "Doctrine\\DBAL\\DBALException",
                "type": "::",
                "args": [{
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOMySql\\Driver"
                    }, {
                        "errorInfo": ["42000", 1064, "You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1"],
                        "__class__": "Doctrine\\DBAL\\Driver\\PDOException"
                    }, "SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)", []]
            }, {
                "file": "/var/www/cloud/lib/private/DB/Connection.php",
                "line": 202,
                "function": "executeQuery",
                "class": "Doctrine\\DBAL\\Connection",
                "type": "->",
                "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)", [], [], null]
            }, {
                "file": "/var/www/cloud/apps/news/lib/Db/FeedMapperV2.php",
                "line": 177,
                "function": "executeQuery",
                "class": "OC\\DB\\Connection",
                "type": "->",
                "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)", []]
            }, {
                "file": "/var/www/cloud/apps/news/lib/Service/FeedServiceV2.php",
                "line": 358,
                "function": "read",
                "class": "OCA\\News\\Db\\FeedMapperV2",
                "type": "->",
                "args": ["anoymouserver", 135, 129183]
            }, {
                "file": "/var/www/cloud/apps/news/lib/Controller/FeedController.php",
                "line": 286,
                "function": "read",
                "class": "OCA\\News\\Service\\FeedServiceV2",
                "type": "->",
                "args": ["anoymouserver", 135, 129183]
            }, {
                "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 169,
                "function": "read",
                "class": "OCA\\News\\Controller\\FeedController",
                "type": "->",
                "args": [135, 129183]
            }, {
                "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 100,
                "function": "executeController",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [{
                        "__class__": "OCA\\News\\Controller\\FeedController"
                    }, "read"]
            }, {
                "file": "/var/www/cloud/lib/private/AppFramework/App.php",
                "line": 152,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [{
                        "__class__": "OCA\\News\\Controller\\FeedController"
                    }, "read"]
            }, {
                "file": "/var/www/cloud/lib/private/Route/Router.php",
                "line": 309,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::",
                "args": ["OCA\\News\\Controller\\FeedController", "read", {
                        "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                    }, {
                        "feedId": "135",
                        "_route": "news.feed.read"
                    }
                ]
            }, {
                "file": "/var/www/cloud/lib/base.php",
                "line": 1008,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->",
                "args": ["/apps/news/feeds/135/read"]
            }, {
                "file": "/var/www/cloud/index.php",
                "line": 37,
                "function": "handleRequest",
                "class": "OC",
                "type": "::",
                "args": []
            }
        ],
        "File": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php",
        "Line": 79,
        "Previous": {
            "Exception": "Doctrine\\DBAL\\Driver\\PDOException",
            "Message": "SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1",
            "Code": "42000",
            "Trace": [{
                    "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php",
                    "line": 911,
                    "function": "query",
                    "class": "Doctrine\\DBAL\\Driver\\PDOConnection",
                    "type": "->",
                    "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)"]
                }, {
                    "file": "/var/www/cloud/lib/private/DB/Connection.php",
                    "line": 202,
                    "function": "executeQuery",
                    "class": "Doctrine\\DBAL\\Connection",
                    "type": "->",
                    "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)", [], [], null]
                }, {
                    "file": "/var/www/cloud/apps/news/lib/Db/FeedMapperV2.php",
                    "line": 177,
                    "function": "executeQuery",
                    "class": "OC\\DB\\Connection",
                    "type": "->",
                    "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)", []]
                }, {
                    "file": "/var/www/cloud/apps/news/lib/Service/FeedServiceV2.php",
                    "line": 358,
                    "function": "read",
                    "class": "OCA\\News\\Db\\FeedMapperV2",
                    "type": "->",
                    "args": ["anoymouserver", 135, 129183]
                }, {
                    "file": "/var/www/cloud/apps/news/lib/Controller/FeedController.php",
                    "line": 286,
                    "function": "read",
                    "class": "OCA\\News\\Service\\FeedServiceV2",
                    "type": "->",
                    "args": ["anoymouserver", 135, 129183]
                }, {
                    "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 169,
                    "function": "read",
                    "class": "OCA\\News\\Controller\\FeedController",
                    "type": "->",
                    "args": [135, 129183]
                }, {
                    "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 100,
                    "function": "executeController",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                    "args": [{
                            "__class__": "OCA\\News\\Controller\\FeedController"
                        }, "read"]
                }, {
                    "file": "/var/www/cloud/lib/private/AppFramework/App.php",
                    "line": 152,
                    "function": "dispatch",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                    "args": [{
                            "__class__": "OCA\\News\\Controller\\FeedController"
                        }, "read"]
                }, {
                    "file": "/var/www/cloud/lib/private/Route/Router.php",
                    "line": 309,
                    "function": "main",
                    "class": "OC\\AppFramework\\App",
                    "type": "::",
                    "args": ["OCA\\News\\Controller\\FeedController", "read", {
                            "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                        }, {
                            "feedId": "135",
                            "_route": "news.feed.read"
                        }
                    ]
                }, {
                    "file": "/var/www/cloud/lib/base.php",
                    "line": 1008,
                    "function": "match",
                    "class": "OC\\Route\\Router",
                    "type": "->",
                    "args": ["/apps/news/feeds/135/read"]
                }, {
                    "file": "/var/www/cloud/index.php",
                    "line": 37,
                    "function": "handleRequest",
                    "class": "OC",
                    "type": "::",
                    "args": []
                }
            ],
            "File": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php",
            "Line": 83,
            "Previous": {
                "Exception": "PDOException",
                "Message": "SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ':userId) AND (feeds.id = :feedId)' at line 1",
                "Code": "42000",
                "Trace": [{
                        "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php",
                        "line": 78,
                        "function": "query",
                        "class": "PDO",
                        "type": "->",
                        "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)"]
                    }, {
                        "file": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php",
                        "line": 911,
                        "function": "query",
                        "class": "Doctrine\\DBAL\\Driver\\PDOConnection",
                        "type": "->",
                        "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)"]
                    }, {
                        "file": "/var/www/cloud/lib/private/DB/Connection.php",
                        "line": 202,
                        "function": "executeQuery",
                        "class": "Doctrine\\DBAL\\Connection",
                        "type": "->",
                        "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)", [], [], null]
                    }, {
                        "file": "/var/www/cloud/apps/news/lib/Db/FeedMapperV2.php",
                        "line": 177,
                        "function": "executeQuery",
                        "class": "OC\\DB\\Connection",
                        "type": "->",
                        "args": ["SELECT `items`.`id` FROM `oc_news_items` `items` INNER JOIN `oc_news_feeds` `feeds` ON items.feed_id = feeds.id WHERE (feeds.user_id = :userId) AND (feeds.id = :feedId)", []]
                    }, {
                        "file": "/var/www/cloud/apps/news/lib/Service/FeedServiceV2.php",
                        "line": 358,
                        "function": "read",
                        "class": "OCA\\News\\Db\\FeedMapperV2",
                        "type": "->",
                        "args": ["anoymouserver", 135, 129183]
                    }, {
                        "file": "/var/www/cloud/apps/news/lib/Controller/FeedController.php",
                        "line": 286,
                        "function": "read",
                        "class": "OCA\\News\\Service\\FeedServiceV2",
                        "type": "->",
                        "args": ["anoymouserver", 135, 129183]
                    }, {
                        "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                        "line": 169,
                        "function": "read",
                        "class": "OCA\\News\\Controller\\FeedController",
                        "type": "->",
                        "args": [135, 129183]
                    }, {
                        "file": "/var/www/cloud/lib/private/AppFramework/Http/Dispatcher.php",
                        "line": 100,
                        "function": "executeController",
                        "class": "OC\\AppFramework\\Http\\Dispatcher",
                        "type": "->",
                        "args": [{
                                "__class__": "OCA\\News\\Controller\\FeedController"
                            }, "read"]
                    }, {
                        "file": "/var/www/cloud/lib/private/AppFramework/App.php",
                        "line": 152,
                        "function": "dispatch",
                        "class": "OC\\AppFramework\\Http\\Dispatcher",
                        "type": "->",
                        "args": [{
                                "__class__": "OCA\\News\\Controller\\FeedController"
                            }, "read"]
                    }, {
                        "file": "/var/www/cloud/lib/private/Route/Router.php",
                        "line": 309,
                        "function": "main",
                        "class": "OC\\AppFramework\\App",
                        "type": "::",
                        "args": ["OCA\\News\\Controller\\FeedController", "read", {
                                "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
                            }, {
                                "feedId": "135",
                                "_route": "news.feed.read"
                            }
                        ]
                    }, {
                        "file": "/var/www/cloud/lib/base.php",
                        "line": 1008,
                        "function": "match",
                        "class": "OC\\Route\\Router",
                        "type": "->",
                        "args": ["/apps/news/feeds/135/read"]
                    }, {
                        "file": "/var/www/cloud/index.php",
                        "line": 37,
                        "function": "handleRequest",
                        "class": "OC",
                        "type": "::",
                        "args": []
                    }
                ],
                "File": "/var/www/cloud/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php",
                "Line": 78
            }
        },
        "CustomMessage": "--"
    },
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:86.0) Gecko/20100101 Firefox/86.0",
    "version": "20.0.8.1",
    "id": "60562c959af29"
}
```
</details>